### PR TITLE
Lb/placements data model

### PR DESCRIPTION
**Pull request by elceebee**
## Context

We have to add some data structures to the code base to enable work on add-a-placement journeys

## Changes proposed in this pull request

- Add subjects, placements and relevant join tables
- Associations on the `Placements::School` and `Placements::Mentor models`
- Update data model documentation to include the table / relationships.
- A bit of placement data in the seeds (this will change when we have real subjects from the Publish API. This is [in flight](https://trello.com/c/Tmr1x5Uo) 

## Guidance to review

Tests should pass -- there are not UI changes. 

## Link to Trello card

https://trello.com/c/1waPU0sp

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/models/placement.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  has_many :placement_subject_joins
+  has_many :subjects, through: :placement_subject_joins
+
+  belongs_to :school, class_name: "Placements::School"
+
+  validates :school, :status, :start_date, :end_date, presence: true
+
+  validate :start_date_before_end_date
+
+  def start_date_before_end_date
```
Comment (JamieCleare2525) [inline app/models/placement.rb:34]: Can't you just use comparison validation?
https://guides.rubyonrails.org/active_record_validations.html#comparison

Comment (elceebee) [reply app/models/placement.rb:34]: Oh duh, I forgot that existed. 
Will change

Comment (elceebee) [reply app/models/placement.rb]: Fixed with commit https://github.com/DFE-Digital/itt-mentor-services/pull/306/commits/22196f33e4c42dfca718e7335c696e78b19787d4

**Comment by elceebee** [comment]
[Data model for placements and related tables](https://trello.com/c/1waPU0sp/142-data-model-for-placements-and-related-tables)

**Review by JamieCleare2525 — APPROVED** [review]
LGTM!

File: docs/data-model.md
Comment (ollietreend) [inline docs/data-model.md:1]: It looks like there's an error in this diagram syntax – it [doesn't render](https://github.com/DFE-Digital/itt-mentor-services/blob/lb/placements-data-model/docs/data-model.md).

I think Mermaid is a bit picky about syntax. I suspect this line might have upset it:

https://github.com/DFE-Digital/itt-mentor-services/blob/22196f33e4c42dfca718e7335c696e78b19787d4/docs/data-model.md?plain=1#L21

Comment (ollietreend) [reply docs/data-model.md:1]: The Mermaid editor is probably a handy way to debug this:

https://mermaid.live/edit#pako:eNqlV9tq4zAQ_ZXBD8sutPsBeSstXbqlkKXpWyDI9iTW1paMLi2hyb_vSJYcXxIn24JzkTJz5sxVykeSyRyTWYLqjrONYtVSADzb9C9mBj7cAsBangM988dmrY3iYgO6kVoxhQyWSa14xdR2mYBUtNSYSZH7jZ6WYBX2NhwBkvcfmaxQw1rJCmqbllwXcDN_uILUGuD0aBDSAIP7R0L1uHv35l7zkmVYoTCB_G_JxSkH_LqOCita3Hd_iY61-2MjT_SS6gs2Kg8wZWIS2CjGBeKYelZIWXa3c2aQQs2UWbmvnU0UeWcLha2cnLE68PGV4PHOFIJVgsxRDudNCcArbmFNVdCw0UBR-vVw8xwrIXWQbWj0SqN645krggeR84wokcoaTIHRfsE0pIgCpEglUzkSDYpfV6SNm4aA1zOXlYxXXzW1UCx7hW8wZ9u-kRCtpirORMucilaNsi7RBevuz6KHO1fyjed4Dtm30HHoAOBT4T1BlhUEuHB15HRvpVWagkHt9ok0tQS_kKheDFteU03gczpqgejr2ZYDRw0K5_iKBg_F3mDeYXLr4CcJjLvtBIOAuGi6dnpkHKH_-cp50Werxmxrl9NDWmYzr0W4PgJhPRjjWDFewsvjwNYTVimVWcHrSSetPuqgVBsmuGaGS7HyvBp3ZbmtpKoLngHTWmbcS8A7N0WsKmIbi3DAtAfqbX4Wsleh5z3tx3dR0OlFD7lO41tCjmuqBXgnBkVsBGoTkqhpVvPMlkxRybZGUiyl2GhS_QndXI3YdPI2_C1G5kQ7HK3p_kGw211f7z7GIZiRh673KybCgR9m4aUKvTEXlILNMXbso72T2x30nOQhTi1se6JG-SOn-NjIabXuDeMCvcj2rGjX4b4bzdUgXsv-h08nFftuKtoBezJzQfzCYMUUXGZkXFHNsB0DN_vTqEzknfY4rjyoko2SttbR1YioA4xvW9nr_naYBr6DYTcmPhBwJOTJUoXv9WEg_ZgCOFoiQ_XkKqlQ0YTO6Wbvp9MyoROYLt6JU6TRw2xpnJk9iTJr5PNWZMnMKItXia3drTD8HUhma1Zq3P8DUs4o_Q

**Review by Nitemaeric — APPROVED** [review]
👍 

I think you might want to add these files to the `CODEOWNERS` file. This way, the top-level Placements models are owned entirely by the Placements team. 😄

File: docs/data-model.md
Comment (elceebee) [reply docs/data-model.md:1]: Fixed with https://github.com/DFE-Digital/itt-mentor-services/pull/306/commits/73137c1e69918a1622a3fba0b7982e3532bd2604

File: docs/data-model.md [hunk 2]
```diff
@@ -16,13 +16,32 @@ There are a few things to bear in mind when reading this:
...
     uuid mentor_id FK
+  }
+  
+  Placement {
+    uuid id PK
     uuid trainee_id FK
     uuid school_id FK
     date start_date
     date end_date
+    enum status
```
Comment (ollietreend) [inline docs/data-model.md:44]:
```suggestion
    enum status "draft or published"
```

File: docs/data-model.md [hunk 3]
```diff
@@ -16,13 +16,32 @@ There are a few things to bear in mind when reading this:
 
 ```mermaid
 erDiagram
-  Placement {
+  Subject {
+    uuid id PK
+    string subject_area "primary or secondary"
+    string name
+    string code "code comes from publish API, but it is not a FK"
```
Comment (ollietreend) [inline docs/data-model.md:23]: "but it is not a FK" – is this true?

I think the API endpoint returns data in an odd format. But my understanding is that the subject "code" is the unique identifier for a subject.

```suggestion
    string code FK "Primary key for subjects in the Teacher Training Courses API"
```

Comment (elceebee) [reply docs/data-model.md:23]: The trouble with that is that code can be null (the Modern languages subject has a code of null) so I think calling it a FK is misleading.

Comment (ollietreend) [reply docs/data-model.md:23]: Fair point. Marking as resolved.

**Review by ollietreend — APPROVED** [review]

**Pull request merged**